### PR TITLE
Pearson correlation among elements of an array

### PR DESCRIPTION
**Pull request by jvdp1**
Related to the functions `cov` and `var`

As usual, everything can/must be discussed.

# Other implementations

* Octave [corr](https://octave.org/doc/v4.2.1/Correlation-and-Regression-Analysis.html)
* Matlab [corr](https://www.mathworks.com/help/stats/corr.html)
* R [cor](https://www.statmethods.net/stats/correlations.html)
* Julia [cor](https://docs.julialang.org/en/v1/stdlib/Statistics/)
* Numpy [corrcoef](https://docs.scipy.org/doc/numpy/reference/generated/numpy.corrcoef.html)

# Specs:

### Description

Returns the Pearson correlation of the elements of `array` along dimension `dim` if the corresponding element in `mask` is `true`.

The Pearson correlation between two rows (or columns), say `x` and `y`, of `array` is defined as:

```
 corr(x, y) = cov(x, y) / sqrt( var(x) * var(y))
```

### Syntax

`result = corr(array, dim [, mask])`

### Arguments

`array`: Shall be a rank-1 or a rank-2 array of type `integer`, `real`, or `complex`.

`dim`: Shall be a scalar of type `integer` with a value in the range from 1 to `n`, where `n` is the rank of `array`.

`mask` (optional): Shall be of type `logical` and either a scalar or an array of the same shape as `array`.

### Return value

If `array` is of rank 1 and of type `real` or `complex`, the result is of type `real` corresponding to the type of `array`.
If `array` is of rank 2 and of type `real` or `complex`, the result is of the same type as `array`.
If `array` is of type `integer`, the result is of type `real(dp)`.

If `array` is of rank 1 and of size larger than 1, a scalar equal to 1 is returned. Otherwise, IEEE `NaN` is returned.
If `array` is of rank 2, a rank-2 array  with the corresponding correlations is returned.

If `mask` is specified, the result is the Pearson correlation of all elements of `array` corresponding to `true` elements of `mask`. If every element of `mask` is `false`, the result is IEEE
 `NaN`.

### Example

```fortran
program demo_corr
    use stdlib_experimental_stats, only: corr
    implicit none
    real :: x(1:6) = [ 1., 2., 3., 4., 5., 6. ]
    real :: y(1:2, 1:3) = reshape([ -1., 40., -3., 4., 10., 6. ], [ 2, 3])
    print *, corr(x, 1)           !returns 1.
    print *, corr(y, 2)           !returns reshape([ 1., -.32480, -.32480, 1. ], [ 2, 3])
end program demo_corr
```

### CODE REVIEW

File: src/tests/stats/test_corr.f90 [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+                            cmplx(0.00000_dp, 1.44065_dp),&
+                            cmplx(1.26401_dp, 0.00000_dp),&
+                            cmplx(0.00000_dp, 0.88833_dp),&
+                            cmplx(1.14352_dp, 0.00000_dp)]
+    complex(dp) :: ds(2,3) = reshape([ cmplx(1._dp, 0._dp),&
+                            cmplx(0._dp, 2._dp),&
+                            cmplx(3._dp, 0._dp),&
+                            cmplx(0._dp, 4._dp),&
+                            cmplx(5._dp, 0._dp),&
+                            cmplx(0._dp, 6._dp)], [2, 3])
```
Comment (certik) [inline src/tests/stats/test_corr.f90:27]: Per #210, you should add `kind=dp` to each `cmplx`.

Comment (jvdp1) [reply src/tests/stats/test_corr.f90]: Fixed

File: doc/specs/stdlib_experimental_stats.md [hunk 2]
```diff
@@ -6,6 +6,54 @@ title: experimental_stats
...
+
+The Pearson correlation between two rows (or columns), say `x` and `y`, of `array` is defined as:
+
+```
+ corr(x, y) = cov(x, y) / sqrt( var(x) * var(y))
+```
+
+### Syntax
+
+`result = [[stdlib_experimental_stats(module):corr(interface)]](array, dim [, mask])`
```
Comment (milancurcic) [inline doc/specs/stdlib_experimental_stats.md:23]: What is the "syntax" of this syntax? What does `[[stdlib_experimental_stats(module):corr(interface)]]` mean? I think this subsection and snippet should show me how to call `corr()`. I don't think this is doing that. Shouldn't it be this:

```suggestion
`result = corr(array, dim[, mask])`
```

File: doc/specs/stdlib_experimental_stats.md [hunk 3]
```diff
@@ -6,6 +6,54 @@ title: experimental_stats
...
+
+`array`: Shall be a rank-1 or a rank-2 array of type `integer`, `real`, or `complex`.
+
+`dim`: Shall be a scalar of type `integer` with a value in the range from 1 to `n`, where `n` is the rank of `array`.
+
+`mask` (optional): Shall be of type `logical` and either a scalar or an array of the same shape as `array`.
+
+### Return value
+
+If `array` is of rank 1 and of type `real` or `complex`, the result is of type `real` corresponding to the type of `array`.
```
Comment (milancurcic) [inline doc/specs/stdlib_experimental_stats.md:35]:
```suggestion
If `array` is of rank 1 and of type `real` or `complex`, the result is of type `real` and has the same kind as `array`.
```

File: doc/specs/stdlib_experimental_stats.md [hunk 4]
```diff
@@ -6,6 +6,54 @@ title: experimental_stats
...
+If `array` is of rank 2 and of type `real` or `complex`, the result is of the same type as `array`.
```
Comment (milancurcic) [inline doc/specs/stdlib_experimental_stats.md:36]:
```suggestion
If `array` is of rank 2 and of type `real` or `complex`, the result is of the same type and kind as `array`.
```

File: doc/specs/stdlib_experimental_stats.md (hunk 2 above)
Comment (milancurcic) [reply doc/specs/stdlib_experimental_stats.md:23]: Oh, is this some special FORD syntax for hyperlinking? If yes, please ignore this comment.

Comment (jvdp1) [reply doc/specs/stdlib_experimental_stats.md:23]: It is for FORD. It creates a link to the module.
See an example for `cov` [here](https://stdlib.fortran-lang.org/page/specs/stdlib_experimental_stats.html#syntax).

**Comment by jvdp1** [comment]
If there are no other objections/comments to this PR, I will then merge this evening.

**Comment by jvdp1** [comment]
Thank you for your inputs. I'll merge.

**Pull request merged**